### PR TITLE
Add seek helper abstract class

### PR DESCRIPTION
**Pull request by garyrussell**
- also handle late arriving time stamp seek requests while processSeeks() is already running.

### CODE REVIEW

**Comment by garyrussell** [comment]
Do not merge - needs more work.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+ * listener subclasses this class, it can easily seek arbitrary topics/partitions without
+ * having to keep track of the callbacks itself.
+ *
+ * @author Gary Russell
+ * @since 2.3
+ *
+ */
+public abstract class AbstractConsumerSeekAware implements ConsumerSeekAware {
+
+	private static final ThreadLocal<ConsumerSeekCallback> callbackForThread = new ThreadLocal<>();
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java]: `static`?
Shouldn't such a thread store be an instance based?

File: spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+	private final Map<TopicPartition, ConsumerSeekCallback> callbacks = new ConcurrentHashMap<>();
+
+	private final Map<TopicPartition, ConsumerSeekCallback> unmodifiable = Collections.unmodifiableMap(this.callbacks);
```
Comment (artembilan) [inline spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java]: I think `unmodifiable` is not so informative name. What is the point to have it as a property?
Seems for me we never do that in other similar places...

File: src/reference/asciidoc/kafka.adoc [hunk 3]
```diff
@@ -2153,6 +2164,35 @@ class Listener implements ConsumerSeekAware {
...
+public class SeekToLastOnIdleListener extends AbstractConsumerSeekAware {
+
+    @KafkaListener(id = "seekOnIdle", topics = "seekOnIdle")
+    public void listen(String in) {
+        ...
+    }
+
+    @Override
+    public void onIdleContainer(Map<org.apache.kafka.common.TopicPartition, Long> assignments,
+            ConsumerSeekCallback callback) {
```
Comment (artembilan) [inline src/reference/asciidoc/kafka.adoc]: Is this argument out of use now?
What is the point to perform a search if we have this argument here?
Or what is the relationship between this argument and whatever is `assignments`?

Thanks for understanding.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java (hunk 1 above)
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java]: Good catch.

File: src/reference/asciidoc/kafka.adoc (hunk 3 above)
Comment (garyrussell) [reply src/reference/asciidoc/kafka.adoc]: Doh; of course; the collection is really only needed when performing seeks from some other arbitrary environment (like in the example above this).

Will polish.

File: spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java (hunk 2 above)
Comment (garyrussell) [reply spring-kafka/src/main/java/org/springframework/kafka/listener/AbstractConsumerSeekAware.java]: OK

**Pull request merged**